Doc revamp and cleanup (#2806)
dfinity/motoko: doc/modules/language-guide/pages/basic-concepts.adoc

:proglang: Motoko
:company-id: DFINITY

> {proglang} is designed for distributed programming with actors. When programming on the Internet Computer, Motoko actors represent IC canister smart contracts.

Author: ```suggestion
{proglang} is designed for distributed programming with actors. When programming on the Internet Computer, Motoko actors represent  {IC} canister smart contracts with Candid interfaces.
```

> {proglang} is designed for distributed programming with actors. When programming on the Internet Computer, Motoko actors represent  {IC} canister smart contracts with Candid interfaces.

Reviewer: I was looking for precisely this sentence, to see where and how the terms "actor" and "canister" are first connected in the text.

I wonder if we could expand this statement slightly to help address @osa1's other comment (now resolved):

> When programming on the Internet Compute in Motoko, each **actor** represents an **{IC} canister smart contract** with a Candid interface, whether written in Motoko, Rust, Wasm or some other language that compiles to Wasm.

> Within Motoko, we use the term *actor* to refer to any canister, authored in any language that deploys to the {IC}.

> The role of Motoko is to make these actors easy to author, and easy to use programmatically, once deployed.

I'm offering those statements as suggestions, but not suggested edits; so feel free to use or not.

BTW, I learned from Norman Ramsey's technical writing notes that using singular forms of nouns is more clear in definitions.

Author: That's nice, I'll use that. I agree with Norman.

Author: ```suggestion
{proglang} is designed for distributed programming with actors. 
When programming on the Internet Computer in Motoko, each **actor** represents an **{IC} canister smart contract** with a Candid interface, whether written in Motoko, Rust, Wasm or some other language that compiles to Wasm.
Within Motoko, we use the term *actor* to refer to any canister, authored in any language that deploys to the {IC}.
The role of Motoko is to make these actors easy to author, and easy to use programmatically, once deployed.
```

{proglang} is designed for distributed programming with actors.

>  When programming on the Internet Computer in Motoko, each **actor** represents an **{IC} canister smart contract** with a Candid interface, whether written in Motoko, Rust, Wasm or some other language that compiles to Wasm. Within Motoko, we use the term *actor* to refer to any canister, authored in any language that deploys to the {IC}. The role of Motoko is to make these actors easy to author, and easy to use programmatically, once deployed.

Reviewer 2: Maybe "once deployed, easy to use programmatically" is easier to read? It's probably just a familiarity thing so feel free to ignore

Reviewer 2: Should the `Motoko`s in this line be `{proglang}`? There are 4 occurrences in this line.

Author: ```suggestion
 When programming on the Internet Computer in {proglang}, each **actor** represents an **{IC} canister smart contract** with a Candid interface, whether written in {proglang}, Rust, Wasm or some other language that compiles to Wasm. Within {proglang}, we use the term *actor* to refer to any canister, authored in any language that deploys to the {IC}. The role of {proglang} is to make these actors easy to author, and easy to use programmatically, once deployed.
```
God forbid  we ever change the name again... Motoko smart blockchain instruction vernacular anyone?

As in the situations above, this function type-checks in all contexts, and when evaluated, traps in all contexts.

[[overview-traps]]
=== Traps due to execution failure

Some errors, such as division by zero, out-of-bounds array indexing, and pattern match failure are not prevented by the type system, but can cause dynamic failures called _traps_.

[source, motoko]
....
1/0; // traps due to division by 0
....

> [source, motoko]

Reviewer: I was surprised to see the text below in this PR.  Is it being truly added, or just moved from some other file?

(Why is it important to have it in this PR, which is mostly about copy editing for terminology?)

Author: It's being added because it was no longer accurate and didn't mention trap. Might as well do somehting useful while rearranging the deck chairs.

As in the situations above, this function type-checks in all contexts, and when evaluated, traps in all contexts.

[[overview-traps]]
=== Traps due to execution failure

Some errors, such as division by zero, out-of-bounds array indexing, and pattern match failure are not prevented by the type system, but can cause dynamic failures called _traps_.

[source, motoko]
....
1/0; // traps due to division by 0
....

[source, motoko]
....
let a = ["hello", "world"];
a[2]; // traps due to out-of-bounds indexing
....

[source, motoko]
....
let true = false; // pattern match failure
....

We say that code _traps_ when its exection causes a _trap_.

Execution of code is aborted at the first trap and makes no further progress.

NOTE: Traps that occur within actor messages are more subtle: they don't abort the entire actor, but prevent that particular message from proceeding, rolling back any yet uncommitted state changes. Other messages on the actor will continue execution.

[[overview-debug-trap]]
=== Explicit traps

Occasionally it can be useful to force an unconditional trap, with a user-defined message.

> The `Debug` library provides a the function `trap(t)` for this purpose, which can be used in any context.

Reviewer 2: "a the"

Author: ```suggestion
The `Debug` library provides the function `trap(t)` for this purpose, which can be used in any context.
```
